proto_proxy.c

```diff
@@ -0,0 +1,2310 @@
+                        // ms <key> <flags>*\r\n
+                        break;
+                    case 'd':
+                        cmd = CMD_MD;
+                        break;
+                    case 'n':
+                        cmd = CMD_MN;
+                        break;
+                    case 'a':
+                        cmd = CMD_MA;
+                        break;
+                    case 'e':
+                        cmd = CMD_ME;
+                        break;
+                }
+            }
+            break;
+        case 3:
+            if (cm[0] == 'g') {
+                if (cm[1] == 'e' && cm[2] == 't') {
+                    cmd = CMD_GET;
+                }
+                if (cm[1] == 'a' && cm[2] == 't') {
+                    cmd = CMD_GAT;
+                }
+            } else if (cm[0] == 's' && cm[1] == 'e' && cm[2] == 't') {
+                cmd = CMD_SET;
+                ret = _process_request_storage(rq, cur, token);
+            } else if (cm[0] == 'a' && cm[1] == 'd' && cm[2] == 'd') {
+                cmd = CMD_ADD;
+                ret = _process_request_storage(rq, cur, token);
+            } else if (cm[0] == 'c' && cm[1] == 'a' && cm[2] == 's') {
+                cmd = CMD_CAS;
+                ret = _process_request_storage(rq, cur, token);
+            }
+            break;
+        case 4:
+            if (cm[0] == 'g' && cm[1] == 'e' && cm[2] == 't' && cm[3] == 's') {
+                cmd = CMD_GETS;
+            } else if (cm[0] == 'i' && cm[1] == 'n' && cm[2] == 'c' && cm[3] == 'r') {
+                cmd = CMD_INCR;
+            } else if (cm[0] == 'd' && cm[1] == 'e' && cm[2] == 'c' && cm[3] == 'r') {
+                cmd = CMD_DECR;
+            }
+            break;
+        case 5:
+            if (strncmp(cm, "touch", 5) == 0) {
+                cmd = CMD_TOUCH;
+            } else if (strncmp(cm, "stats", 5) == 0) {
+                cmd = CMD_STATS;
+                // :key() should give the stats sub-command?
+            }
+            break;
+        case 6:
+            if (strncmp(cm, "delete", 6) == 0) {
+                cmd = CMD_DELETE;
+            } else if (strncmp(cm, "append", 6) == 0) {
+                cmd = CMD_APPEND;
+                ret = _process_request_storage(rq, cur, token);
+            }
+            break;
+        case 7:
+            if (strncmp(cm, "replace", 7) == 0) {
+                cmd = CMD_REPLACE;
+                ret = _process_request_storage(rq, cur, token);
+            } else if (strncmp(cm, "prepend", 7) == 0) {
+                cmd = CMD_PREPEND;
+                ret = _process_request_storage(rq, cur, token);
+            } else if (strncmp(cm, "version", 7) == 0) {
+                cmd = CMD_VERSION;
+            }
+            break;
+    }
+
+    // TODO: check ret and fail if needed.
+    if (ret == -1) {
+
+    }
+    // TODO: check if cmd unfound? need special code?
```

Comment:
I haven't fully decided yet. I'm currently redoing that code as the proxy is supposed to allow hooks on specific commands or command states (ie; only fire a proxy on MISS, only on SET/DELETE's, pass-through STATS and VERSION, etc). But also in previous proxies for other systems I've allowed arbitrary commands, which ends up allowing users to invent commands which do compilated work. I'm not 100% decided on if and/or how I should allow this yet. At first it will probably require strict parsing and then I would relax it later.

Reply:
Interesting.

My main concern for interoperability would be to make it possible to unambiguously parse any custom requests, e.g. with a meta-eval command (which indicates whether or not the command is associated with a payload (or multiple payloads?), similar to the meta set and `v`(?) flag?)

Reply:
Probably yeah. Haven't given it any thought yet; just don't want to close the door. One thing that'll happen sooner is I'm reserving a meta flag (`R` or something) as a "router hint", which operates like a second opaque value but clients should ignore completely. This can be useful in a lot of ways:

- Shift the "prefix route" style hints into R flag. Then we don't have to modify the client string to avoid storing the route path on the backend servers. Also slightly better performance.
- On the same line allows preserving the route info when hitting backend servers; normally you'd want to strip the prefix but you'd lose where it comes from when observing requests directly on the backend.
- Since it's an arbitrary opaque routers could even modify it if they wanted. Might even reserve two of such flags, one which should only be used by proxies and one which is client/either.